Paper: Will Millennials Ever Get Married? (Downey)

AllenDowney opened the pull request:
I've drafted my paper in LaTeX and would like to upload the LaTeX source and PDF version.  Is that enough to get the review process started?  Then I can work on converting to rst and getting it to build.

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/30

AllenDowney commented:
This now contains an rst version that builds, with lots of little things that need fixing.

AllenDowney commented:
Ok, I think this is ready for review.

dalippa commented:
Hi @AllenDowney. I was asked to join the review of your paper. Below is my feedback:

1. I suggest using numbered footnotes, references, and appendices to maximize the flow of ideas.

2. Paper needs proper citations.

3. Why are you usuing only data from female respondents and not both male and female? I find that point particularly interesting, and could use a little more discussion.

4. I noticed some grammar and punctuation mistakes scattered throughout the paper.

5. Citations for the survival analysis might be useful.

6. The survival analysis section is interesting, but it is a bit convoluted, alternating between the code and describing survival analysis.

7. The projection section seems to contradict the title of the paper (prediction of social trends is futile, and the title of the paper hints that the paper will answer that question).

8. In the future work section, you may wish to consider culture as a factor.

cchoirat commented:
Hi @AllenDowney:

Here are some comments about your contribution:

- Abstract: "Of women born in the 1980s, 60% are married by age 32; of women born in 1990s, 13% are married by age 22."  This comparison is not very helpful without the discussion of p. 4.
- typo p. 1: women’shealth.
- It would be helpful to provide a literature reference to Kaplan-Meir besides/instead of Wikipedia.
- Could you please compare your approach and code with other survival analysis tools (e.g., https://pypi.python.org/pypi/lifelines/)?

AllenDowney commented:
@dalippa  Thanks for these comments.  I've made some changes in the paper to address the ones I could.  To answer one of your questions: the only reason we have not analyzed the male data is that we have not got to it yet.  It will be interesting to look at but, because it doesn't not go back as far, it is likely to be less informative.

AllenDowney commented:
@cchoirat Thanks.  I made some changes in the abstract to address your point (and fixed the typo).  I removed the reference to Wikipedia.  I have a brief mention of and reference to lifelines.

jseabold commented:
A few meta-comments.

1. I echo the pep-8 comment.
2. You may want to define millennials somewhere.
3. You might want to introduce the link to your code repo. earlier.
4. Echo the citations comment. For the software packages as well.
5. It might make sense to talk about pandas Series in terms of labeled indices rather than as mappings. I don't know if that will be natural to everyone. E.g., 'the series contains with hazard rates with times as the row labels index' or something.

AllenDowney commented:
@jseabold Thanks!  Since some of these are repeats, I should explain why I have not made some requested changes:

1) About PEP-8, the code base here is from Think Stats, which follows the Google style guide for historical reasons.  It is all PEP-8 except camel-case function/method names, and I think that's more trouble than it's worth to change.  PEP-8 is great, but it's not dogma.

2 and 3) Yes.

4) It makes sense to me to put full citations into the references section.  But for things that are just links, what's the point?  I know there are style guides for turning URLs into standard-looking citations, but I think they're silly.

5) I dunno.  I think it's natural to talk about a function as a map from its domain to its range.  And a Series is a mapping type as well as a sequence type.

jseabold commented:
On Tue, Jun 30, 2015 at 3:03 PM, Allen Downey <notifications@github.com>
wrote:

> @jseabold <https://github.com/jseabold> Thanks! Since some of these are
> repeats, I should explain why I have not made some requested changes:
>
> 1) About PEP-8, the code base here is from Think Stats, which follows the
> Google style guide for historical reasons. It is all PEP-8 except
> camel-case function/method names, and I think that's more trouble than it's
> worth to change. PEP-8 is great, but it's not dogma.
>
Sure. We've had this conversation before. It's just never going to be not
odd to see CamelCase method and function names to me. It's different from
most of the scientific python code people are used to reading. Some style
is better than no style though.


> 2 and 3) Yes.
>
> 4) It makes sense to me to put full citations into the references section.
> But for things that are just links, what's the point? I know there are
> style guides for turning URLs into standard-looking citations, but I think
> they're silly.
>
When possible, I think they should be more than links. If there's a
citation page and/or a paper for the software package you should cite it.

> 5) I dunno. I think it's natural to talk about a function as a map from
> its domain to its range. And a Series is a mapping type as well as a
> sequence type.
>
Your call.

jseabold commented:
You might also consider mentioning version numbers of software used or adding a requirements.txt to the repo. Things are more stable these days but probably wouldn't hurt.

katyhuff commented:
@AllenDowney & @jseabold Fun fact: It's entirely the fault of the google style guide that our reviewer guidelines aren't more hard-lined about PEP8. Google Style onward, folks.

jseabold commented:
Do they have conflicting style guides?

https://google.github.io/styleguide/pyguide.html#Naming

AllenDowney commented:
Google's external style guide is consistent with PEP-8; the internal style
guide is not (or at least was not in 2009).

On Tue, Jun 30, 2015 at 9:10 PM, Skipper Seabold <notifications@github.com>
wrote:

> Do they have conflicting style guides?
>
> https://google.github.io/styleguide/pyguide.html#Naming
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/123#issuecomment-117383637>
> .
>

katyhuff commented:
And, the spacing is still not pep8 compliant (https://code.google.com/p/soc/wiki/PythonStyleGuide#Python_Style_Rules).

dalippa commented:
Hi @AllenDowney. Most of my comments were not addressed, neither in the form of a comment nor a change in the paper:

1. I am not seeing numbered footnotes and appendicies per my previous comment. This really does break the flow of the paper.

2. Referenced web sites still need proper citation style.

3. Comment addressed in github.

4. Some paragraphs are 1 sentence long. Combine into a single paragraph, or expound upon the ideas. Consider active voice over passive voice. The sentence "So we can compute with resolution of one month each respondent’s age at interview, age, and age at first marriage, agemarry" is grammatically incorrect. The two pargraphs on cohorts should be combined.

5. Still missing citations for survival analysis.

6. The survival analysis section is much better, but referencing code variables (such as "complete") still impedes readability. Consider using an alternate means since the variables are lowercase, yet start a sentence.

7. The introductory paragraph to the "Projection" section still seems to directly contradict the title/message of the paper (prediction of social trends is futile, and the title of the paper hints that the paper will answer that question). Consider removing the senteince "Predicitng these kinds of social trends...". The second sentence seems sufficient, if combined with the second paragraph.

8. In the future work section, you may wish to consider culture as a factor.

AllenDowney commented:
@dalippa Thank you for all of your comments.  I didn't mean to leave them hanging, but I wasn't sure that a point-by-point response was needed.

1) I am not seeing numbered footnotes and appendicies per my previous comment. This really does break the flow of the paper.
2) Referenced web sites still need proper citation style.

I think there are several acceptable styles for citations.  The one I chose is the one I prefer.


4) Some paragraphs are 1 sentence long. Combine into a single paragraph, or expound upon the ideas. Consider active voice over passive voice. The sentence "So we can compute with resolution of one month each respondent’s age at interview, age, and age at first marriage, agemarry" is grammatically incorrect. The two pargraphs on cohorts should be combined.

I don't think there's anything wrong with these.

5) Still missing citations for survival analysis.

In an earlier version I included a link to the Wikipedia page on Kaplan-Meier estimation.  Another reviewer suggested that it was unnecessary, and I agreed that it would be easy for a reader to find more resources without a link.

6) The survival analysis section is much better, but referencing code variables (such as "complete") still impedes readability. Consider using an alternate means since the variables are lowercase, yet start a sentence.

Variable names are typeset in a different font.  Some of the typesetting was missing, but I think I've got all of them now.

7) The introductory paragraph to the "Projection" section still seems to directly contradict the title/message of the paper (prediction of social trends is futile, and the title of the paper hints that the paper will answer that question). Consider removing the senteince "Predicitng these kinds of social trends...". The second sentence seems sufficient, if combined with the second paragraph.

The paper says what I want to say about this.

8) In the future work section, you may wish to consider culture as a factor.

This paper doesn't address causal factors, and I don't have plans for future work that does.

I'm sorry if it seemed like I wasn't addressing your comments, but in cases where I disagree with a reviewer, I am not sure it is helpful to debate.  It seems like this system fosters more conflict than necessary.

dalippa commented:
@AllenDowney

My comments were suggestions, and definitely were not intended to engender any conflict. I think some debate is healthy, however.

1) I was mostly concerned with the readability of the paper. It's an interesting subject, and when reading it, I found some of the side-notes distracting. Feel free to disregard.

2) I was mostly concerned with having a consistent citation style: MLA, APA, Chicago, etc. Feel free to disregard.

4) I'll concede the point on the two cohort paragraphs, but sentences should have a complete thought; "So we can compute with resolution of one month each respondent’s age at interview, age, and age at first marriage, agemarry" is still grammatically incorrect. Perhaps, "Therefore, we can compute with resolution of one month each respondent’s age at interview, age, and age at first marriage, agemarry".

5) OK.

6) OK.

7) OK.

8) On this one, I substantially disagree. If country of origin and region are considered salient, culture most definitely should be considered for mention in Future Work. I am of the opinion that the future work section is not just to address what the author(s) might do in the future, but also for readers and other researchers as well.

josef-pkt commented on papers/allen_downey/allen_downey.rst:
```diff
@@ -0,0 +1,591 @@
+
+For each value of ``t`` we look up ``ended``, which is the number of people
+married at ``t``, and ``censored``, which is the number of people interviewed at
+t. The hazard function at ``t`` is the ratio of ``ended`` and ``at_risk``.
```
`censored` is the number of women interviewed at age t **that are not married**

phrasing for `ended` may be ambigous: AFAI it's women that get married (for the frst time) at t (a flow variable, not the stock of women that are married)

AllenDowney replied:
@josef-pkt Thanks, will do.

AllenDowney commented:
@dalippa 

4) Sorry, I did not understand until now that you were objecting to beginning the sentence with "So".  The consensus among grammarians is that this is fine: http://www.quickanddirtytips.com/education/grammar/can-i-start-a-sentence-with-a-conjunction

8) I think future work should be about what I intend to do, not what anyone could do.  The second option would be very long.

dalippa commented:
@AllenDowney 

4) OK.

8) I'm OK with leaving it out, but it does seem like a logical next step like the other population segments you mentioned (http://cs.stanford.edu/people/widom/paper-writing.html#future)

katyhuff commented:
@sbenthall @dalippa @cchoirat @jseabold : Can you confirm that the changes satisfy your recommendations sufficient for inclusion in the proceedings?

katyhuff commented:
I see your opinion on full references for code is "what's the point". However, I'd like to direct you to our guidelines which require code citations where possible. From the point of view of the SciPy organizers involved in that recommendation the point is that this practice recognizes that code is a type of research product and is important to the flourishing of our community. Furthermore, internet links are not as persistent as citations to particular dated versions of those code and data products, even if the ultimate citation involves a URL. 

Note that many of the tools in the scipy stack request particular citations on their webpages or in the CITATION file of their source code. Additionally, the SciPy community has a page with common software and how to cite it here: (http://www.scipy.org/citing.html). Currently, none of the following mentioned tools are cited appropriately in your paper:

NumPy
Pandas
lifelines
(also, the cdc data http://www.cdc.gov/nchs/nsfg.htm)

jseabold commented:
On Thu, Jul 2, 2015 at 4:14 PM, Katy Huff <notifications@github.com> wrote:

> I see your opinion on full references for code is "what's the point".
> However, I'd like to direct you to our guidelines which require code
> citations where possible. From the point of view of the SciPy organizers
> involved in that recommendation the point is that this practice recognizes
> that code is a type of research product and is important to the flourishing
> of our community. Furthermore, internet links are not as persistent as
> citations to particular dated versions of those code and data products,
> even if the ultimate citation involves a URL.
>
Well put.

This is the sentiment I had hoped to get across. I'd hope that proper
citations also help legitimize the writing of software as research/academic
output within the broader academic community not just this one.

My only other comment was to add version numbers to help with
reproducibility, given software like pandas is often a fast moving target.

dalippa commented:
@jseabold @AllenDowney @katyhuff My primary issue at the moment is the lack of proper citations. Once that's done, I would be willing to give it a :+1:

AllenDowney commented:
I have made the requested changes in the bibliography.  However, I seem to have broken something along the way: the references appear at the end of the paper, as expected, but not in a section labelled References.  I can't see how the changes I made causes this effect.  Maybe someone who knows more about rst can help me.

katyhuff commented:
1) Thanks!!!
2) I'll give it a look sometime today or tomorrow.

On Fri, Jul 3, 2015 at 9:56 AM, Allen Downey <notifications@github.com>
wrote:

> I have made the requested changes in the bibliography. However, I seem to
> have broken something along the way: the references appear at the end of
> the paper, as expected, but not in a section labelled References. I can't
> see how the changes I made causes this effect. Maybe someone who knows more
> about rst can help me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/123#issuecomment-118392344>
> .
>



-- 
http://katyhuff.github.com

katyhuff commented:
Aha, this is because you now use numbers rather than something alphanumeric. Numbered links like that [1]_ [2]_ and so on are the footnote syntax in rst.  It's a shame that rst can't handle numbered references. It will be fixed if you use a syntax like the following instead. Sorry about that:

```
.. [Wan2014] Wendy Wang and Kim etc.

.. [Wal2011] Stefan van der Walt etc.

.. [McK2010] Wes McKinney etc.

.. [Dav2015] Cameron Davidson-Pilon etc.

.. [Dow2014] Allen Downey, etc.
```

AllenDowney commented:
Done.  Thanks for the debugging help!

On Sat, Jul 4, 2015 at 5:06 PM, Katy Huff <notifications@github.com> wrote:

> Aha, this is because you now use numbers rather than something
> alphanumeric. Numbered links like that [1]_ [2]_ and so on are the footnote
> syntax in rst. It's a shame that rst can't handle numbered references. It
> will be fixed if you use a syntax like the following instead. Sorry about
> that:
>
> .. [Wan2014] Wendy Wang and Kim etc.
>
> .. [Wal2011] Stefan van der Walt etc.
>
> .. [McK2010] Wes McKinney etc.
>
> .. [Dav2015] Cameron Davidson-Pilon etc.
>
> .. [Dow2014] Allen Downey, etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/123#issuecomment-118555725>
> .
>
